### PR TITLE
Anchor links

### PR DESCRIPTION
**Pull request by suhjainadobe**
If the URL has anchor links, the browser is not scrolling automatically to that particular element.

Resolves:
 https://jira.corp.adobe.com/browse/MWPW-131971
https://jira.corp.adobe.com/browse/MWPW-133331

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/suhjain/blade-runner-2049s-oscar-winning-texturing-workflow-at-framestore#the-solar-farm
- After: https://anchorlinks--milo--suhjainadobe.hlx.page/drafts/suhjain/blade-runner-2049s-oscar-winning-texturing-workflow-at-framestore#the-solar-farm

### CODE REVIEW

File: libs/scripts/scripts.js
Comment (mokimo) [inline libs/scripts/scripts.js:159]: The scripts are defined on a per-consumer basis - so every consumer would need to redefine this, I don't think this is the right place to call this function, otherwise each consumer would need to repeat this

File: libs/utils/utils.js
Comment (mokimo) [inline libs/utils/utils.js:946]: Why is that needed? That seems like an evil hack

**Review by mokimo — COMMENTED** [review]
The linter has a lot of complaints make sure to fix those and check if your vscode/lint setup is running correctly

File: libs/utils/utils.js
Comment (mokimo) [inline libs/utils/utils.js:951]:
```suggestion
  if (hash) {
    const elementId = hash.slice(1);
    const targetElement = document.querySelector(`#${elementId}:not(.dialog-modal)`);
    if (targetElement) {
      const position = targetElement.getBoundingClientRect();
      let bufferHeight = document.querySelector('.global-navigation')?.offsetHeight || 0
      window.scrollTo({
        top: position.top - bufferHeight,
        behavior: 'smooth',
      });
```

Comment (suhjainadobe) [reply libs/utils/utils.js]: done

File: libs/scripts/scripts.js
Comment (mokimo) [inline libs/scripts/scripts.js:17]: Unused import

File: libs/utils/utils.js
Comment (mokimo) [inline libs/utils/utils.js:900]: I'm a bit concerned about the timing of this, but I'll leave it up to the milo core team to have the final call on this

File: test/utils/utils.test.js
Comment (mokimo) [inline test/utils/utils.test.js:434]: Should this test setup be equal to the other ones setting a bool?
```
      window.scrollTo = () => {
        scrollToCalled = true;
      };
```

**Review by mokimo — APPROVED** [review]
Some last comments and would like for the milo core team to have some input on this

File: libs/utils/utils.js
Comment (suhjainadobe) [reply libs/utils/utils.js:900]: Sorry @mokimo , this will not work, as load eventListener does not get trigger here and Safari is not working as expected. Will have to add it in script.js. Looking for other ways to handle this problem. scrollToHashedElement() is fine, it is working as expected but figuring out how to handle load eventListener without calling it in script.js

**Comment by suhjainadobe** [comment]
Closing to check some changes of the load eventListener.

**Comment by suhjainadobe** [comment]
@mokimo and @aishwaryamathuria : Could you please review it again.

File: libs/utils/utils.js
Comment (chrischrischris) [inline libs/utils/utils.js:895]: nit: only set this if `currentHash` has a value

**Review by chrischrischris — CHANGES_REQUESTED** [review]
This is breaking modals - see:
https://anchorlinks--milo--suhjainadobe.hlx.page/drafts/suhjain/blade-runner-2049s-oscar-winning-texturing-workflow-at-framestore#playmodal

The hash keeps getting added and removed, causing modal to flicker / reload.  Closing the modal opens it up right away again.

File: libs/utils/utils.js
Comment (suhjainadobe) [reply libs/utils/utils.js:897]: done

**Comment by suhjainadobe** [comment]
@chrischrischris : The modal flickering is resolved. Please review again.

**Review by seanchoi-dev — APPROVED** [review]

**Comment by suhjainadobe** [comment]
> On safari, it scrolls past the anchor title

@chrischrischris : Could you please let me know when is it not working? For me it is scrolling down before the title only.

**Comment by chrischrischris** [comment]
Probably out of scope for this, but should a Header that's inside a tab that's not currently displayed activate that tab and scroll to it?

**Comment by ryanmparrish** [comment]
@suhjainadobe  Just an FYI, I tested this page w/ a recently added block `marquee-anchors` which has some anchor link similarities. I don't see any issues w/ compatibility. 
Here's the test page https://anchorlinks--milo--suhjainadobe.hlx.page/drafts/rparrish/marquee-nav#how-to-block

**Comment by suhjainadobe** [comment]
> @suhjainadobe Just an FYI, I tested this page w/ a recently added block `marquee-anchors` which has some anchor link similarities. I don't see any issues w/ compatibility.

@ryanmparrish : Here too I could see the issue intermittently. It was scrolling past the title in Safari.

File: libs/utils/utils.js
Comment (suhjainadobe) [reply libs/utils/utils.js]: done

**Comment by suhjainadobe** [comment]
> Probably out of scope for this, but should a Header that's inside a tab that's not currently displayed activate that tab and scroll to it?

If the header is not present in the active tab, then the page will not scroll down and activate the tab. We yet are not supporting this feature.

**Comment by suhjainadobe** [comment]
@chrischrischris , @ryanmparrish : I have made some changes in the style.css file which accommodates the default scrolling in Safari. It also resolved [MWPW-133331](https://jira.corp.adobe.com/browse/MWPW-133331).

File: libs/styles/styles.css
Comment (mokimo) [inline libs/styles/styles.css:89]: I don't think a hardcoded value will work long term

Comment (suhjainadobe) [reply libs/styles/styles.css]: done

File: libs/styles/styles.css
Comment (mokimo) [inline libs/styles/styles.css:93]: That's definitely better than a static value, though one Issue I forsee is when things are on top of the navigation.

There are already such requirements https://github.com/orgs/adobecom/discussions/1166

I don't think we necessarily need to fix this right now but we should keep it in mind, otherwise the scroll-location might be off in those cases.

Comment (mokimo) [reply libs/styles/styles.css:93]: Can you explain a bit why the `scroll-margin-top` is needed and the top-offset isn't enough?
```
  window.scrollTo({
    top: topOffset - bufferHeight,
    behavior: 'smooth',
  });
```

Comment (suhjainadobe) [reply libs/styles/styles.css:93]: On safari, the scrollbar is automatically initiated even before the scrollToHashedElement is initiated, so when it first scrolls down, the nav bar hides the title. Also to resolve [MWPW-133331](https://jira.corp.adobe.com/browse/MWPW-133331), we need this CSS property.

**Review by chrischrischris — APPROVED** [review]

**Review by chrischrischris — COMMENTED** [review]
One nit: combine header selectors to save some file size:
```
h1, h2, h3, h4, h5, h6 {
  scroll-margin-top: var(--feds-totalheight-nav);
}
```

**Pull request merged**
